### PR TITLE
[device_info_platform_interface][device_info] nnbd

### PR DESCRIPTION
**Pull request by cyanglaz**
Migrate to NNBD

### CODE REVIEW

**Comment by blasten** [comment]
We would need to opt out the driver tests in the PR for `device_info` until `integration_test` and `vm_service` are migrated to null safety. 

FYI @dnfield

File: packages/device_info/device_info/example/pubspec.yaml [hunk 1]
```diff
@@ -12,7 +12,11 @@ dev_dependencies:
...
     path: ../../../integration_test
-  pedantic: ^1.8.0
+  pedantic: ^1.10.0-nullsafety.1
 
 flutter:
   uses-material-design: true
+
+environment:
+  sdk: '>=2.10.0-56.0.dev <3.0.0'
+
```
Comment (blasten) [inline packages/device_info/device_info/example/pubspec.yaml:22]: nit: EOF

**Comment by dnfield** [comment]
@bkonyi  what's the plan for null safe vm_service?

File: packages/device_info/device_info/example/lib/main.dart [hunk 2]
```diff
@@ -59,13 +59,13 @@ class _MyAppState extends State<MyApp> {
...
-      'version.incremental': build.version.incremental,
-      'version.codename': build.version.codename,
-      'version.baseOS': build.version.baseOS,
+      'version.securityPatch': build.version?.securityPatch,
+      'version.sdkInt': build.version?.sdkInt,
+      'version.release': build.version?.release,
+      'version.previewSdkInt': build.version?.previewSdkInt,
+      'version.incremental': build.version?.incremental,
+      'version.codename': build.version?.codename,
+      'version.baseOS': build.version?.baseOS,
```
Comment (blasten) [inline packages/device_info/device_info/example/lib/main.dart:68]: should `version` be non-nullable?  In what situation should it be `null`?

File: packages/device_info/device_info/example/lib/main.dart [hunk 3]
```diff
@@ -98,11 +98,11 @@ class _MyAppState extends State<MyApp> {
...
-      'utsname.sysname:': data.utsname.sysname,
-      'utsname.nodename:': data.utsname.nodename,
-      'utsname.release:': data.utsname.release,
-      'utsname.version:': data.utsname.version,
-      'utsname.machine:': data.utsname.machine,
+      'utsname.sysname:': data.utsname?.sysname,
+      'utsname.nodename:': data.utsname?.nodename,
+      'utsname.release:': data.utsname?.release,
+      'utsname.version:': data.utsname?.version,
+      'utsname.machine:': data.utsname?.machine,
```
Comment (blasten) [inline packages/device_info/device_info/example/lib/main.dart:105]: similar comment about nullability.

File: packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart [hunk 4]
```diff
@@ -21,57 +21,57 @@ class AndroidDeviceInfo {
...
-    List<String> systemFeatures,
+    List<String> systemFeatures = const [],
   })  : supported32BitAbis = List<String>.unmodifiable(supported32BitAbis),
         supported64BitAbis = List<String>.unmodifiable(supported64BitAbis),
         supportedAbis = List<String>.unmodifiable(supportedAbis),
         systemFeatures = List<String>.unmodifiable(systemFeatures);
 
   /// Android operating system version values derived from `android.os.Build.VERSION`.
-  final AndroidBuildVersion version;
+  final AndroidBuildVersion? version;
```
Comment (blasten) [inline packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:38]: For this one, and the ones below. It doesn't seem like they are set to `null` in the platform code: https://github.com/flutter/plugins/blob/42b33ec8ed3265607df5accf45bc68aa9e4e5592/packages/device_info/device_info/android/src/main/java/io/flutter/plugins/deviceinfo/MethodCallHandlerImpl.java#L39

Comment (blasten) [inline packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:38]: It seems like only `baseOS`, `previewSdkInt`, and `securityPatch` should be `null`:
https://github.com/flutter/plugins/blob/42b33ec8ed3265607df5accf45bc68aa9e4e5592/packages/device_info/device_info/android/src/main/java/io/flutter/plugins/deviceinfo/MethodCallHandlerImpl.java#L39

File: packages/device_info/device_info/example/lib/main.dart (hunk 2 above)
Comment (cyanglaz) [reply packages/device_info/device_info/example/lib/main.dart:68]: All these values in `AndroidDeviceInfo` came from method channel as a map, so I feel like it is probably safest to make them nullable? As we don't have control over the result from method channel. 
The other way I can think of is to provide a default value to some of the variables so they can be non-nullable. I'm not sure if that's a better approach tho.

**Review by cyanglaz — COMMENTED** [review]
@blasten Updated per review comments, PTAL

File: packages/device_info/device_info/pubspec.yaml [hunk 5]
```diff
@@ -19,14 +22,15 @@ flutter:
 dependencies:
   flutter:
     sdk: flutter
-  device_info_platform_interface: ^1.0.0
+  device_info_platform_interface:
+    path: ../device_info_platform_interface
```
Comment (blasten) [inline packages/device_info/device_info/pubspec.yaml:26]: We are going with git dependencies, but feel free to change it after this PR is merged.

**Review by blasten — APPROVED** [review]
LGTM

File: packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart [hunk 6]
```diff
@@ -8,28 +8,28 @@
...
-    List<String> supported32BitAbis,
-    List<String> supported64BitAbis,
-    List<String> supportedAbis,
-    this.tags,
-    this.type,
-    this.isPhysicalDevice,
-    this.androidId,
-    List<String> systemFeatures,
-  })  : supported32BitAbis = List<String>.unmodifiable(supported32BitAbis),
+    required this.version,
```
Comment (mehmetf) [inline packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:11]: We have forked this package internally and my migration followed a different path. I am not a fan of making all these fields `required` for two reasons:

- Constructing the object (for tests) becomes painful. Consider an app test that only needs to set `manufacturer`. They now have to initialize all the other fields in their tests or use the obscure Map constructor.
- Semantically, these fields are not required to exist. They *can be* empty. To reflect that, we are initializing them to empty string. As an app developer, I would much rather have a signal that a field is nullable rather than having it be an empty string which does not provide compile-time safety.

In my migration, I marked all these fields as nullable. However, this failed tests because gallery uses this code and assumes they are not nullable.

Comment (blasten) [reply packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:11]: > Semantically, these fields are not required to exist. They can be empty. To reflect that, we are initializing them to empty string.

If it's referring to usage of `Map`, this is a limitation of the implementation. 

When looking at this, we were trying to answer the question of what is a valid `AndroidDeviceInfo` based on the information available from the platform. If something changes in future versions of Android, then that is a breaking change to the API exposed to Dart.

In my opinion, if everything is nullable by default, then the API becomes some sort of "best-effort".   "we think `version` is available, but we can't guarantee that". 

This was the rationale behind it.  That said, could you point to the cases where testing was painful?

Comment (mehmetf) [reply packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:11]: > we were trying to answer the question of what is a valid AndroidDeviceInfo based on the information available from the platform.

That means these fields are *not* supposed to be nullable. In that case, it would be far better to construct them via map values like so:

```dart
   someField: map[key]!,
   // rather than
   // someField: map[key] ?? '',
```

The code, today, reads like any one of these values can be non-existent and that's OK.

The test pain is a secondary reason. If these are truly non-nullable values, you can always expose a `forTest` constructor that effectively does what your map constructor is doing.

Comment (blasten) [reply packages/device_info/device_info_platform_interface/lib/model/android_device_info.dart:11]: `someField: map[key]!` makes sense to me. We should fix it.  This is an area where I think Pigeon could help, but that is a different discussion 😃
